drizzlepac/hlautils/astrometric_utils.py

```diff
@@ -871,187 +804,193 @@ def build_self_reference(filename, clean_wcs=False):
-    wcsout.wcs.crpix = np.array([crpix1,crpix2])
+    wcsout.wcs.crval = crval2.copy()
+    wcsout.wcs.crpix = crpix.copy()
     wcsout.naxis1 = naxis1
     wcsout.naxis2 = naxis2
-    wcsout.wcs.cd = buildRotMatrix(orientat)*[-1,1]*pscale/3600.0
-    # Synchronize updates with PyWCS/WCSLIB objects
+    wcsout.wcs.cd = buildRotMatrix(orientat) * [-1, 1] * pscale / 3600.0
+    # Synchronize updates with astropy.wcs objects
     wcsout.wcs.set()
     wcsout.setPscale()
     wcsout.setOrient()
-    wcsout.wcs.ctype = ['RA---TAN','DEC--TAN']
+    wcsout.wcs.ctype = ['RA---TAN', 'DEC--TAN']
 
     return wcsout
 
-def within_footprint(img, wcs, x, y):
-    """Determine whether input x,y fall in the science area of the image.
+
+def within_footprint(img, wcsobj, x, y):
+    """Determine whether input x, y fall in the science area of the image.
 
     Parameters
-    -----------
+    ----------
     img : ndarray
-        ndarray of image where non-science areas are marked with value of NaN
+        ndarray of image where non-science areas are marked with value of NaN.
 
-    wcs : obj
-        HSTWCS or WCS object with naxis terms defined
+    wcsobj : `stwcs.wcsutil.HSTWCS`
+        HSTWCS or WCS object with naxis terms defined.
 
-    x,y : arrays
-        arrays of x,y positions for sources to be checked
+    x, y : ndarray
+        arrays of x, y positions for sources to be checked.
 
     Returns
     -------
-    x,y : arrays
+    x, y : ndarray
         New arrays which have been trimmed of all sources that fall outside
         the science areas of the image
 
     """
     # start with limits of WCS shape
-    if hasattr(wcs, 'naxis1'):
-        naxis1 = wcs.naxis1
-        naxis2 = wcs.naxis2
-    elif hasattr(wcs, 'pixel_shape'):
-        naxis1, naxis2 = wcs.pixel_shape
-    else:
-        naxis1 = wcs._naxis1
-        naxis2 = wcs._naxis2
-    maskx = np.bitwise_or(x<0, x>naxis1)
-    masky = np.bitwise_or(y<0, y>naxis2)
-    mask = ~np.bitwise_or(maskx,masky)
-    x = x[mask]
-    y = y[mask]
-
-    # Now, confirm that these points fall within actual science area of WCS
-    img_mask = create_image_footprint(img, wcs, border=1.0)
-    inmask = np.where(img_mask[y.astype(np.int32),x.astype(np.int32)])[0]
+    #if hasattr(wcs, 'naxis1'):
+        #naxis1 = wcs.naxis1
+        #naxis2 = wcs.naxis2
+    #elif hasattr(wcs, 'pixel_shape'):
+        #naxis1, naxis2 = wcs.pixel_shape
+    #else:
+        #naxis1 = wcs._naxis1
+        #naxis2 = wcs._naxis2
+    #maskx = np.bitwise_or(x < 0, x > naxis1)
+    #masky = np.bitwise_or(y < 0, y > naxis2)
+    #mask = ~np.bitwise_or(maskx, masky)
+    #x = x[mask]
+    #y = y[mask]
+
+    ## Now, confirm that these points fall within actual science area of WCS
+    #img_mask = create_image_footprint(img, wcs, border=1.0)
+    #inmask = np.where(img_mask[y.astype(np.int32), x.astype(np.int32)])[0]
+
+    sky = wcs.pixel_to_world(x, y, 1)
+    inmask = wcs.footprint_contains(sky)
```

Comment:
`wcs.footprint_contains`  is a new addition to astropy v 3.1. I am adding it as a suggestion here as it simplifies the code.
You may need to loop over all chips though in order to reproduce the current functionality.